firedrake/interpolation.py

```diff
@@ -130,16 +170,27 @@ def interpolate(self, *function, output=None, transpose=False):
 
         if self.nargs:
             function, = function
-            if transpose:
-                mul = assembled_interpolator.handle.multTranspose
-                V = self.arguments[0].function_space()
+            if not hasattr(function, "dat"):
+                raise ValueError("The expression had arguments: we therefore need to be given a Function (not an expression) to interpolate!")
+            if not self.vom_onto_other_vom:
+                if transpose:
+                    mul = assembled_interpolator.handle.multTranspose
+                    V = self.arguments[0].function_space()
+                else:
+                    mul = assembled_interpolator.handle.mult
+                    V = self.V
+                result = output or firedrake.Function(V)
+                with function.dat.vec_ro as x, result.dat.vec_wo as out:
+                    mul(x, out)
+                return result
             else:
-                mul = assembled_interpolator.handle.mult
-                V = self.V
-            result = output or firedrake.Function(V)
-            with function.dat.vec_ro as x, result.dat.vec_wo as out:
-                mul(x, out)
-            return result
+                if transpose:
+                    V = self.arguments[0].function_space()
+                else:
+                    V = self.V
+                result = output or firedrake.Function(V)
+                assembled_interpolator(result.dat, function.dat, transpose)
```

Comment:
Can we somehow avoid having `make_interpolator` return `mom_to_other_vom` by making fake `handle.mult` and `handle.multTranspose` available, and remove `if` here?

Reply:
I think with some suitable edits (perhaps introducing a dummy `handle` class, such as `VomOntoVomDummyMat` with `mult` and `multTranspose` properties) here
https://github.com/firedrakeproject/firedrake/blob/89659fc8ade94aab66cedc53a7fecbafa4a94f4d/firedrake/interpolation.py#L286-L295
I should be able to.

Style question: should I define the class inside the `make_interpolator` (if that's even allowed) or would it be better to put it close to `VomOntoVomCallable`? I presume the latter?

Reply:
An issue: I'd need to modify `VomOntoVomCallable` to work with PETSc `vec`s instead of `dat`s. I don't know how I would get the correct MPI type in that case. Currently I have the below:
https://github.com/firedrakeproject/firedrake/blob/89659fc8ade94aab66cedc53a7fecbafa4a94f4d/firedrake/interpolation.py#L860
Are you aware of a way to get the appropriate MPI type in this situation? If not, your comment might not be actionable at the moment...

Reply:
It would be better to put the `handle` class close to `VomOntoVomCallable`.

MPI data type does not change as we do not make any change to the underlying data. I think your question is how to get numpy array from vec to use the `petsc4py` interface of `sf.bcast` and friends. I thought we could use `vec.getArray()`, but I realised `vec` does not include data on halo, which you need, so what I suggested is probably not as simple as I had thought.

Reply:
Oh so we don't generally include halos when interpolating? It's possible that I could make the `input_ordering_sf` not include halos in the leaves (only the leaves include halos). A halo exchange would then force halo dofs to be updated. Since we already aren't including halos I presume we don't force a halo exchange after interpolation?

Reply:
Yes, that would be more consistent with interpolations on regular meshes. `input_ordering_sf` I think is standard, so we probably don't want to change it directly. Maybe we can use this: PetscSFCreateEmbeddedLeafSF https://gitlab.com/petsc/petsc/-/blob/main/src/binding/petsc4py/src/petsc4py/PETSc/SF.pyx#L323.

Reply:
> Yes, that would be more consistent with interpolations on regular meshes. `input_ordering_sf` I think is standard, so we probably don't want to change it directly. Maybe we can use this: PetscSFCreateEmbeddedLeafSF https://gitlab.com/petsc/petsc/-/blob/main/src/binding/petsc4py/src/petsc4py/PETSc/SF.pyx#L323.

I think using `PetscSFCreateEmbeddedLeafSF` would require a-priori knowledge of the rank distribution of the leaves (recall that input ordering vertices are roots and our VOM's mesh vertices are leaves). I think, for simplicity, I'll have two properties: `input_ordering_sf`, which includes halo cells, and `input_ordering_no_halos_sf`, which does not.